src/err.cc

```diff
@@ -0,0 +1,157 @@
+const e::exctype e::ProcessLookupError(PyExc_ProcessLookupError);
+#if LIBPY_HAVE_RECURSION_ERROR
+const e::exctype e::RecursionError(PyExc_RecursionError);
+#endif
+const e::exctype e::ReferenceError(PyExc_ReferenceError);
+const e::exctype e::RuntimeError(PyExc_RuntimeError);
+const e::exctype e::SyntaxError(PyExc_SyntaxError);
+const e::exctype e::SystemError(PyExc_SystemError);
+const e::exctype e::TimeoutError(PyExc_TimeoutError);
+const e::exctype e::SystemExit(PyExc_SystemExit);
+const e::exctype e::TypeError(PyExc_TypeError);
+const e::exctype e::ValueError(PyExc_ValueError);
+const e::exctype e::ZeroDivisionError(PyExc_ZeroDivisionError);
+#if MS_WINDOWS
+const e::exctype e::WindowsError(PyExc_WindowsError);
+#endif
+
+// OSError aliases
+const py::err::exctype EnvironmentError(PyExc_EnvironmentError);
+const py::err::exctype IOError(PyExc_IOError);
+
+py::err::object::object() : py::object() {}
+
+py::err::object::object(PyObject *pob) : py::object(pob) {
+    exception_check();
+}
+
+py::err::object::object(const py::object &pob) : py::object(pob) {
+    exception_check();
+}
+
+py::err::object::object(const py::err::object &cpfrom) :
+    py::object(cpfrom.ob) {}
+
+py::err::object::object(py::err::object &&mvfrom) noexcept :
+    py::object(mvfrom.ob) {
+    mvfrom.ob = nullptr;
+}
+
+void py::err::object::exception_check() {
+    if (ob && !PyExceptionClass_Check(Py_TYPE(ob))) {
+        ob = nullptr;
+        if (!PyErr_Occurred()) {
+            PyErr_SetString(TypeError,
+                            "cannot make py::err::object from non exception");
+        }
+    }
+}
+
+py::tmpref<py::object> py::err::object::traceback() const {
+    if (!is_nonnull()) {
+        pyutils::failed_null_check();
+        return nullptr;
+    }
+    return PyException_GetTraceback(ob);
+}
+
+int py::err::object::traceback(py::object tb) {
+    if (!is_nonnull()) {
+        pyutils::failed_null_check();
+        return -1;
+    }
+    return PyException_SetTraceback(ob, tb);
+}
+
+py::tmpref<py::object> py::err::object::context() const {
+    if (!is_nonnull()) {
+        pyutils::failed_null_check();
+        return nullptr;
+    }
+    return PyException_GetContext(ob);
+}
+
+void py::err::object::context(py::object tb) {
+    if (!is_nonnull()) {
+        pyutils::failed_null_check();
+        return;
+    }
+    PyException_SetContext(ob, tb);
+}
+
+py::tmpref<py::object> py::err::object::cause() const {
+    if (!is_nonnull()) {
+        pyutils::failed_null_check();
+        return nullptr;
+    }
+    return PyException_GetCause(ob);
+}
+
+void py::err::object::cause(py::object cs) {
+    if (!is_nonnull()) {
+        pyutils::failed_null_check();
+        return;
+    }
+    PyException_SetCause(ob, cs);
+}
+
+py::err::msgbuilder::msgbuilder(py::err::exctype type) :
+    type(type), fire(true) {}
+py::err::msgbuilder::msgbuilder(msgbuilder &&mvfrom) :
+    std::stringstream(std::forward<msgbuilder>(mvfrom)), type(mvfrom.type) {
```

Comment:
I want to make sure that dispatches to the super class's move constructor instead of the copy constructor so I want to preserve the rvalue type through the method dispatch. To see this in action here is a small example:

```c++
#include <iostream>

struct C {
    C() {}

    C(const C&) {
        std::cout << "copy\n";
    }

    C(C&&) {
        std::cout << "move\n";
    }
};

void f(C&& c) {
    C copy(c);
    C move(std::forward<C>(c));
}

int main() {
    f(C{});
}
```

running this prints:
```
copy
move
```